File: api/app/schemas/progress.py
from enum import Enum
from typing import Optional, List
from pydantic import BaseModel
from datetime import datetime


class PhaseType(str, Enum):
    """High-level execution phases"""
    INITIALIZATION = "initialization"
    AGENT_EXECUTION = "agent_execution"
    COMPLETION = "completion"


class StepType(str, Enum):
    """Individual steps within phases"""
    # Initialization steps
    CONNECTING_SERVER = "connecting_server"
    INITIALIZING_EXECUTION = "initializing_execution"  
    CREATING_REQUEST = "creating_request"
    
    # Agent execution steps
    AGENT_STARTING = "agent_starting"
    AGENT_WORKING = "agent_working"
    AGENT_FINISHING = "agent_finishing"
    
    # Completion steps
    PROCESSING_RESULTS = "processing_results"
    CREATING_PR = "creating_pr"


class StepStatus(str, Enum):
    """Status of individual steps"""
    IN_PROGRESS = "in_progress"   # Step is currently being executed
    COMPLETED = "completed"       # Step completed successfully
    FAILED = "failed"            # Step failed with error


class AgentContext(BaseModel):
    """Context information for agent steps"""
    agent_id: Optional[str] = None            # Unique ID for this agent instance
    agent_ide: Optional[str] = None           # "cursor", "windsurf", etc.
    agent_role: Optional[str] = None          # "Coder", "Tester", etc.
    agent_model: Optional[str] = None         # "claude-sonnet-4", etc.


class ProgressEvent(BaseModel):
    """Single progress event - for internal use"""
    task_id: str
    step_id: str
    status: StepStatus
    phase_type: PhaseType
    step_type: StepType
    agent_context: Optional[AgentContext] = None
    error_message: Optional[str] = None
    timestamp: datetime


class WebSocketProgressMessage(BaseModel):
    """Message sent over WebSocket to frontend"""
    type: str = "progress"
    task_id: str
    step_id: str
    status: StepStatus
    phase: PhaseType
    step: StepType
    agent_context: Optional[AgentContext] = None
    error_message: Optional[str] = None
    timestamp: datetime


class PreGeneratedStep(BaseModel):
    """A pre-generated step that will be executed"""
    step_id: str  # Unique identifier for this step
    phase: PhaseType
    step: StepType
    agent_context: Optional[AgentContext] = None
    step_order: int  # Order within the entire task execution
    description: Optional[str] = None  # Human-readable description


class TaskStepsPlan(BaseModel):
    """Complete pre-generated plan of all steps for a task"""
    task_id: str
    steps: List[PreGeneratedStep]
    total_steps: int
    estimated_duration_seconds: Optional[int] = None

File: tauri-macos/src/types/progress.ts
// Backend schema types - matching the backend exactly

export enum PhaseType {
  INITIALIZATION = "initialization",
  AGENT_EXECUTION = "agent_execution", 
  COMPLETION = "completion"
}

export enum StepType {
  // Initialization steps
  CONNECTING_SERVER = "connecting_server",
  INITIALIZING_EXECUTION = "initializing_execution",
  CREATING_REQUEST = "creating_request",
  
  // Agent execution steps
  AGENT_STARTING = "agent_starting",
  AGENT_WORKING = "agent_working", 
  AGENT_FINISHING = "agent_finishing",
  
  // Completion steps
  PROCESSING_RESULTS = "processing_results",
  CREATING_PR = "creating_pr"
}

export enum StepStatus {
  IN_PROGRESS = "in_progress",
  COMPLETED = "completed",
  FAILED = "failed"
}

export interface AgentContext {
  agent_id?: string;
  agent_ide?: string;
  agent_role?: string;
  agent_model?: string;
}

export interface PreGeneratedStep {
  step_id: string;
  phase: PhaseType;
  step: StepType;
  agent_context?: AgentContext;
  step_order: number;
  description?: string;
}

export interface TaskStepsPlan {
  task_id: string;
  steps: PreGeneratedStep[];
  total_steps: number;
  estimated_duration_seconds?: number;
}

export interface WebSocketProgressMessage {
  type: string;
  task_id: string;
  step_id: string;
  status: StepStatus;
  phase: PhaseType;
  step: StepType;
  agent_context?: AgentContext;
  error_message?: string;
  timestamp: string;
}

// Frontend-specific types for display

export interface StepDisplay {
  id: string;
  title: string;
  status: StepStatus;
  duration?: string;
  completed: boolean;
  inProgress: boolean;
  failed: boolean;
  agent_context?: AgentContext;
  startTime?: Date;
  endTime?: Date;
}

export interface PhaseDisplay {
  name: string;
  icon: 'rocket' | 'check' | 'test' | 'initialization' | 'code' | 'completion';
  steps: StepDisplay[];
  agent_context?: AgentContext;
}

export interface TaskProgress {
  phases: PhaseDisplay[];
  currentStepId?: string;
  totalSteps: number;
  completedSteps: number;
  failedSteps: number;
}